hail/src/main/scala/is/hail/utils/FlipbookIterator.scala

```diff
@@ -84,6 +86,65 @@ object FlipbookIterator {
     StagingIterator(sm)
 
   def empty[A] = StagingIterator(StateMachine.terminal[A])
+
+  def multiZipJoin[A: ClassTag](
+    its: IndexedSeq[FlipbookIterator[A]],
+    default: A,
+    ord: (A, A) => Int
+  ): FlipbookIterator[Array[A]] = {
+    val staging = its.map(_.toStagingIterator)
+    val sm = new StateMachine[Array[A]] {
+      private val len = its.length
+      var value: Array[A] = Array.fill(len)(default)
+      var isValid = true
+      var buf = new ArrayBuffer[Int]
+      def advance() {
+        buf.clear()
+        var smallest = -1
+        var i = 0
+        while (i < value.length) {
```

Comment:
You can completely avoid this loop if you return just the non-zero values and their indices.

Reply:
So, we'd change the return type to something like `FlipbookIterator[is.hail.utils.ArrayBuilder[(Int, A)]]` and then just call `clear` on every iteration?